### PR TITLE
UserProfile Rework

### PR DESCRIPTION
**Pull request by AliRKat**
Naming is UserProfile Rework since Unity had the functionality prior to this but as it can be seen, old class which doesn't follow the naming convention is already deprecated and new interface and class is added.

Scenario tests contains all of the tests that are currently written.
UserProfileTests can be updated, if any suggestion please mention.
Haven't mentioned "SetUserProperties" method of CountlyConfiguration within changelog, it can also be added.
TestUtility update contains a validator and extractor specifically for UserProfile requests
Main functionality can be found under UserProfile.cs

ps: haven't added comments within the interface since the class itself contains the comments, those also can be added

gl it's kind of a big boi 😄

### CODE REVIEW

**Comment by AliRKat** [comment]
[[Unity] User Property Rework](https://trello.com/c/FsNAATbX/1828-unity-user-property-rework)

File: Assets/Countly/Plugins/CountlySDK/Interfaces/IUserProfileModule.cs [hunk 1]
```diff
@@ -0,0 +1,20 @@
+using System.Collections;
+using System.Collections.Generic;
+using UnityEngine;
+
```
Comment (arifBurakDemiray) [inline Assets/Countly/Plugins/CountlySDK/Interfaces/IUserProfileModule.cs:4]: comments here bra

Comment (arifBurakDemiray) [reply Assets/Countly/Plugins/CountlySDK/Interfaces/IUserProfileModule.cs:4]: done bra

File: Assets/Countly/Plugins/CountlySDK/Services/UserProfile.cs [hunk 2]
```diff
@@ -0,0 +1,682 @@
...
+
+    internal Dictionary<string, object> CustomDataProperties = new Dictionary<string, object>();
+    internal Dictionary<string, JObject> CustomMods = new Dictionary<string, JObject>();
+    private readonly Countly cly;
+    private readonly CountlyConfiguration config;
+    private readonly CountlyUtils utils;
+    internal readonly RequestCountlyHelper requestHelper;
+    internal readonly EventCountlyService eventService;
+    internal UserProfile(Countly countly, CountlyConfiguration configuration, CountlyLogHelper logHelper, RequestCountlyHelper requestCountlyHelper, CountlyUtils countlyUtils, ConsentCountlyService consentService, EventCountlyService events) : base(configuration, logHelper, consentService)
+    {
```
Comment (arifBurakDemiray) [inline Assets/Countly/Plugins/CountlySDK/Services/UserProfile.cs:45]: Oh is not the baseservice holding reference to the cly obj. That is interestimg

Comment (arifBurakDemiray) [reply Assets/Countly/Plugins/CountlySDK/Services/UserProfile.cs:45]: also conf

Comment (arifBurakDemiray) [reply Assets/Countly/Plugins/CountlySDK/Services/UserProfile.cs:45]: It is ok btw it is just that android and probably ios not like this

File: Assets/Countly/Plugins/CountlySDK/Services/UserProfile.cs [hunk 3]
```diff
@@ -0,0 +1,682 @@
...
+            return;
+        }
+
+        Log.Info($"[UserProfile][IncrementInternal] key: [{key}] and value: [{value}]");
+        ModifyCustomData(key, value, "$inc");
+    }
+
+    private void MaxInternal(string key, double value)
+    {
+        if (!cly.IsSDKInitialized) {
```
Comment (arifBurakDemiray) [inline Assets/Countly/Plugins/CountlySDK/Services/UserProfile.cs:216]: could not we move this is init check to the validation function, it looks like all the other methods are also checking this

File: Assets/Countly/Plugins/CountlySDK/Services/UserProfile.cs [hunk 4]
```diff
@@ -0,0 +1,682 @@
...
+    }
+
+    private void PushUniqueInternal(string key, string value)
+    {
+        if (!cly.IsSDKInitialized) {
+            Log.Warning("[UserProfile][PushUniqueInternal], Countly.Instance.Init() must be called before PushUniqueInternal");
+            return;
+        }
+
+        if (!ValidateConsentAndKey(key)) {
```
Comment (arifBurakDemiray) [inline Assets/Countly/Plugins/CountlySDK/Services/UserProfile.cs:296]: this lacks function nmae

File: Assets/Countly/Plugins/CountlySDK/Services/UserProfile.cs (hunk 3 above)
Comment (AliRKat) [reply Assets/Countly/Plugins/CountlySDK/Services/UserProfile.cs:216]: yea makes sense, let me do it

File: Assets/Countly/Plugins/CountlySDK/Services/UserProfile.cs [hunk 5]
```diff
@@ -0,0 +1,682 @@
...
+            }
+
+            if (Organization != null) {
+                if (Organization == "") {
+                    json.Add(ORG_KEY, null);
+                } else {
+                    json.Add(ORG_KEY, Organization);
+                }
+            }
+
```
Comment (arifBurakDemiray) [inline Assets/Countly/Plugins/CountlySDK/Services/UserProfile.cs:575]: this string checks are looking same. MB we can extract them to a function
like
it takes value, jobj and key. And adds it to the jobj if conditions met

File: Assets/Countly/Plugins/CountlySDK/Services/UserProfile.cs (hunk 2 above)
Comment (AliRKat) [reply Assets/Countly/Plugins/CountlySDK/Services/UserProfile.cs:45]: it doesn't have one for the cly that is correct. it does have one for config but i guess if we would refactor these we would do all of them at the same time for all 'Services' to make sure all is right

File: Assets/Countly/Plugins/CountlySDK/Services/UserProfile.cs [hunk 6]
```diff
@@ -0,0 +1,682 @@
...
+                    json.Add(BYEAR_KEY, BirthYear);
+                } else {
+                    json.Add(BYEAR_KEY, null);
+                }
+            }
+
+            JObject ob = null;
+
+            if (CustomDataProperties.Count > 0) {
+                utils.TruncateSegmentationValues(CustomDataProperties, config.GetMaxSegmentationValues(), "[UserProfile][ConvertToJSON]", Log);
```
Comment (arifBurakDemiray) [inline Assets/Countly/Plugins/CountlySDK/Services/UserProfile.cs:611]: btw log structure is not same as [SDK dev guide](https://support.countly.com/hc/en-us/articles/360037753291-SDK-development-guide#01H822TC8FN7N2AHT7C6TJKR95)
afaik it should be
[CLASS] functionName, LOG

Comment (AliRKat) [reply Assets/Countly/Plugins/CountlySDK/Services/UserProfile.cs:611]: oh. didn't enjoy seeing this one 😄

**Review by arifBurakDemiray — CHANGES_REQUESTED** [review]
I will look into to the test classes afterwards. They are long 😅

File: Assets/Tests/PlayModeTests/Scenarios/UP_UserProfile.cs [hunk 7]
```diff
@@ -0,0 +1,410 @@
...
+            };
+
+            return expectedUserDetails;
+        }
+
+        // UserProfile calls in Countly.Instance.UserProfile
+        // We initialize SDK without consent requirement, check the RQ and EQ sizes, then call public methods
+        // Methods we call should record nothing in the RQ and EQ
+        [Test]
+        public void UP_200_CNR_A()
```
Comment (arifBurakDemiray) [inline Assets/Tests/PlayModeTests/Scenarios/UP_UserProfile.cs:100]: maybe expected output might be added
image

File: Assets/Tests/PlayModeTests/Scenarios/UP_UserProfile.cs [hunk 8]
```diff
@@ -0,0 +1,410 @@
...
+        // UserProfile calls in Countly.Instance.UserProfile
+        // We initialize SDK with consent requirement and give consents, check the RQ and EQ sizes, then call public methods
+        // Methods we call should record nothing in the RQ and EQ
+        [Test]
+        public void UP_201_CR_CG_A()
+        {
+            Consents[] consent = new Consents[] { Consents.Crashes, Consents.Events, Consents.Clicks, Consents.StarRating, Consents.Views, Consents.Users, Consents.Push, Consents.RemoteConfig, Consents.Location, Consents.Feedback, Consents.Sessions };
+            CountlyConfiguration config = TestUtility.CreateBaseConfigConsent(consent);
+            Countly.Instance.Init(config);
+            TestUtility.ValidateRQEQSize(Countly.Instance, 2, 0);
```
Comment (arifBurakDemiray) [inline Assets/Tests/PlayModeTests/Scenarios/UP_UserProfile.cs:121]: I assume second request is consent request?

Comment (arifBurakDemiray) [reply Assets/Tests/PlayModeTests/Scenarios/UP_UserProfile.cs:121]: - "// Methods we call should record nothing in the RQ and EQ"

Comment (arifBurakDemiray) [reply Assets/Tests/PlayModeTests/Scenarios/UP_UserProfile.cs:121]: please also updat expecteds

File: Assets/Tests/PlayModeTests/Scenarios/UP_UserProfile.cs [hunk 9]
```diff
@@ -0,0 +1,410 @@
...
+        // UserProfile calls in Countly.Instance.UserProfile
+        // We initialize SDK with consent requirement and give no consent, check the RQ and EQ sizes, then call public methods
+        // Methods we call should record nothing in the RQ and EQ
+        [Test]
+        public void UP_202_CR_CNG_A()
+        {
+            Consents[] consent = new Consents[] { };
+            CountlyConfiguration config = TestUtility.CreateBaseConfigConsent(consent);
+            Countly.Instance.Init(config);
+            TestUtility.ValidateRQEQSize(Countly.Instance, 2, 0);
```
Comment (arifBurakDemiray) [inline Assets/Tests/PlayModeTests/Scenarios/UP_UserProfile.cs:137]: if there is no consent what are those two?
location and consent right?

File: Assets/Tests/PlayModeTests/Scenarios/UP_UserProfile.cs [hunk 10]
```diff
@@ -0,0 +1,410 @@
...
+
+            SendUserData();
+            SendUserProperty();
+            TestUtility.ValidateRQEQSize(Countly.Instance, 2, 0);
+        }
+
+        // RecordEventAsync with UserProfile changes
+        // We record events without consent requirement, between events we record user profile data
+        // With each event, if user data information is recorded, it should flush EQ and send user profile data
```
Comment (arifBurakDemiray) [inline Assets/Tests/PlayModeTests/Scenarios/UP_UserProfile.cs:146]: please also add expecteds to all

File: Assets/Tests/PlayModeTests/Scenarios/UP_UserProfile.cs [hunk 11]
```diff
@@ -0,0 +1,410 @@
...
+            Countly cly = Countly.Instance;
+            TestUtility.ValidateRQEQSize(cly, 1, 0);
+
+            _ = Countly.Instance.Events.RecordEventAsync("BasicEventA");
+            _ = Countly.Instance.Events.RecordEventAsync("BasicEventB");
+            SendSameData();
+            _ = Countly.Instance.Events.RecordEventAsync("BasicEventC");
+            
+            // Extract and validate user_details requests
+            TestUtility.ValidateRQEQSize(cly, 3, 1);
```
Comment (arifBurakDemiray) [inline Assets/Tests/PlayModeTests/Scenarios/UP_UserProfile.cs:162]: please also comment here just simple ones like
// 1 begin session + 1 event req (A,B)+ 1 user profile

Comment (arifBurakDemiray) [reply Assets/Tests/PlayModeTests/Scenarios/UP_UserProfile.cs:162]: to clarify expecteds

**Review by arifBurakDemiray — COMMENTED** [review]
PLease also add tests for individual function calls like

setProperty_nullKey
setProperty_nullValue
mininum_MAX_INT_VALUE
push_null

File: Assets/Tests/PlayModeTests/Scenarios/UP_UserProfile.cs (hunk 9 above)
Comment (AliRKat) [reply Assets/Tests/PlayModeTests/Scenarios/UP_UserProfile.cs:137]: yup sir

File: Assets/Tests/PlayModeTests/Scenarios/UP_UserProfile.cs (hunk 7 above)
Comment (AliRKat) [reply Assets/Tests/PlayModeTests/Scenarios/UP_UserProfile.cs:100]: yes cool addition adding now

File: Assets/Tests/PlayModeTests/Scenarios/UP_UserProfile.cs (hunk 8 above)
Comment (AliRKat) [reply Assets/Tests/PlayModeTests/Scenarios/UP_UserProfile.cs:121]: yes and agreed. adding

File: Assets/Tests/PlayModeTests/Scenarios/UP_UserProfile.cs (hunk 11 above)
Comment (AliRKat) [reply Assets/Tests/PlayModeTests/Scenarios/UP_UserProfile.cs:162]: adding

File: Assets/Tests/PlayModeTests/UserProfileTests.cs [hunk 12]
```diff
@@ -449,6 +449,312 @@ public void SessionTimerElapseWithRecordedValue_M_CR_CG()
...
+        [Test]
+        public void IncrementBy_BadKey_CNR()
+        {
+            Countly cly = BaseInit();
+
+            cly.UserProfile.IncrementBy(null, 10);
+            cly.UserProfile.Save();
+            TestUtility.ValidateRQEQSize(cly, 0, 0);
+
+            cly.UserProfile.IncrementBy(" ", 10);
```
Comment (arifBurakDemiray) [inline Assets/Tests/PlayModeTests/UserProfileTests.cs:500]: ow whitespace is not valid?

**Review by arifBurakDemiray — APPROVED** [review]
Ganbatte A-dono <3 
Just one thing for @turtledreams 
Does whitespace is a valid string for values?

File: Assets/Tests/PlayModeTests/UserProfileTests.cs (hunk 12 above)
Comment (AliRKat) [reply Assets/Tests/PlayModeTests/UserProfileTests.cs:500]: whitespace and null values are ignored btw

**Review by turtledreams — CHANGES_REQUESTED** [review]
nice work! 👏 
you expose setdata for named only and setuserproperties for custom+named to the user? you should just provide setuserproperties to the user. no need to complicate things. u can use setdataintenal internally if you want to process things.
you log 'count' for things, instead log each key/value pairs and if possible their types. I think you created a methods like that before.
you commented 'pull' creates and array if value does not exists?
if you crop picture url u should log it
bool is unsupported?
instead of checking the whitespace you can trim all values initially and if empty stting provided you can log about the value will be erased at server
you should not convert empty strings to null

File: Assets/Countly/Plugins/CountlySDK/Services/AbstractBaseService.cs [hunk 13]
```diff
@@ -105,6 +124,11 @@ protected string TrimValue(string fieldName, string v)
                 v = v.Substring(0, _configuration.GetMaxValueSize());
             }
 
+            if (v == " " || v == "") {
```
Comment (turtledreams) [inline Assets/Countly/Plugins/CountlySDK/Services/AbstractBaseService.cs:127]: this TrimValue now has 2 jobs. lets not do that. just use the Trim() method in C# to trim the string values you receive for user properties. and just check empty space normally and print warning.

**Review by turtledreams — CHANGES_REQUESTED** [review]

File: Assets/Countly/Plugins/CountlySDK/Services/AbstractBaseService.cs (hunk 13 above)
Comment (AliRKat) [reply Assets/Countly/Plugins/CountlySDK/Services/AbstractBaseService.cs]: updated

File: CHANGELOG.md [hunk 14]
```diff
@@ -1,6 +1,20 @@
 ## XX.XX.X
 * Added support for string key and Array/List value to all user given segmentations.
 * Added "SetID(string newDeviceId)" function in "Countly.Instance.Device" for managing device id changes.
+* Added "Countly.Instance.UserProfile" with following calls and functionality
+  * "Increment" for incrementing custom property value by 1
+  * "IncrementBy" for incrementing custom property value by provided value.
+  * "Max" for saving maximal value between existing and provided.
+  * "Min" for saving minimal value between existing and provided.
```
Comment (turtledreams) [inline CHANGELOG.md:8]: this

File: CHANGELOG.md [hunk 15]
```diff
@@ -1,6 +1,20 @@
...
+  * "Multiply" for multiplying custom property value by provided value.
+  * "Pull" for removing value from array.
+  * "Push" for inserting value to array which can have duplicates.
+  * "PushUnique" for inserting value to array of unique values.
+  * "Save" for sending provided values to server.
+  * "SetOnce" for setting value if it does not exist.
+  * "SetData" for setting user profile data.
```
Comment (turtledreams) [inline CHANGELOG.md:15]: this

File: Assets/Countly/Plugins/CountlySDK/Services/UserProfile.cs [hunk 16]
```diff
@@ -0,0 +1,609 @@
...
+    }
+
+    private void SetPropertyInternal(string key, object value)
+    {
+        if (!ValidateConsentAndKey(key, "SetPropertyInternal")) {
+            return;
+        }
+
+        if (utils.IsNullEmptyOrWhitespace(key)) {
+            Log.Info($"[UserProfile] SetPropertyInternal, Provided key is null, empty or whitespace. Ignoring.");
```
Comment (turtledreams) [inline Assets/Countly/Plugins/CountlySDK/Services/UserProfile.cs:413]: warning

File: Assets/Countly/Plugins/CountlySDK/Services/UserProfile.cs [hunk 17]
```diff
@@ -0,0 +1,609 @@
...
+            }
+
+            if (value == null || !(value is double || value is int || value is string || value is bool)) {
+                Log.Warning($"[UserProfile] ModifyCustomData, Provided 'value' is {(value == null ? "null" : "an unsupported type")}");
+                return;
+            }
+
+            string trimmedKey = key.Trim();
+            if (key != trimmedKey) {
+                Log.Warning($"[UserProfile] ModifyCustomData, Provided key: [{key}], for value: [{value}] has been trimmed.");
```
Comment (turtledreams) [inline Assets/Countly/Plugins/CountlySDK/Services/UserProfile.cs:448]: verbose

File: Assets/Countly/Plugins/CountlySDK/Services/UserProfile.cs [hunk 18]
```diff
@@ -0,0 +1,609 @@
...
+            }
+
+            string truncatedKey = TrimKey(trimmedKey);
+            object truncatedValue;
+
+            if (value is string stringValue) {
+                string trimmed = stringValue.Trim();
+                if (stringValue != trimmed) {
+                    Log.Warning($"[UserProfile] ModifyCustomData, Provided value: [{value}], for key: [{key}] has been trimmed. Current value: [{trimmed}]");
```
Comment (turtledreams) [inline Assets/Countly/Plugins/CountlySDK/Services/UserProfile.cs:457]: verbose

**Review by turtledreams — CHANGES_REQUESTED** [review]

**Review by turtledreams — APPROVED** [review]

**Pull request merged**